File: examples/Delphi/Virtual Listbox/Example/EXVLBU.pas
{$J+} {Writable constants}

unit Exvlbu;

interface

uses
  SysUtils, Windows, Messages, Classes, Graphics, Controls,
  Forms, Dialogs, OvcBase, OvcVlb, StdCtrls, Buttons;

type
  TForm1 = class(TForm)
    OvcVirtualListbox1: TOvcVirtualListBox;
    Label1: TLabel;
    Label2: TLabel;
    OvcController1: TOvcController;
    Label3: TLabel;
    Label4: TLabel;
    BitBtn1: TButton;
    procedure OvcVirtualListbox1GetItem(Sender: TObject; Index: Longint;
      var ItemString: string);
    procedure OvcVirtualListbox1GetItemColor(Sender: TObject;
      Index: Longint; var FG, BG: TColor);
    procedure OvcVirtualListbox1Click(Sender: TObject);
    procedure OvcVirtualListbox1IsSelected(Sender: TObject; Index: Longint;
      var Selected: Boolean);
    procedure OvcVirtualListbox1Select(Sender: TObject; Index: Longint;
      Selected: Boolean);
    procedure OvcVirtualListbox1TopIndexChanged(Sender: TObject;
      NewTopIndex: Longint);
    procedure OvcVirtualListBox2GetItem(Sender: TObject; Index: Integer;
      var ItemString: String);
    procedure btnCloseClick(Sender: TObject);
  private
    { Private declarations }

    {maintain selection status of the first 500 items}
    Select : array[0..500] of Boolean;
  public
    { Public declarations }
  end;

var
  Form1: TForm1;

implementation

{$R *.DFM}

procedure TForm1.OvcVirtualListbox1GetItem(Sender: TObject; Index: Longint;
  var ItemString: {$IFDEF Win32} string {$ELSE} OpenString {$ENDIF});
begin
  {cheat and just return the index number as the item string}
  ItemString := Format('Item # %d', [Index]);
end;

procedure TForm1.OvcVirtualListbox1GetItemColor(Sender: TObject;
  Index: Longint; var FG, BG: TColor);
begin
  if Index <> OvcVirtualListbox1.ItemIndex then begin
    if Index mod 13 = 0 then
      FG := clRed
    else if Index mod 7 = 0 then
      FG := clGreen
    else if Odd(Index) then
      FG := clBlue;
  end;
end;

procedure TForm1.OvcVirtualListbox1Click(Sender: TObject);
var
  I : LongInt;
begin
  I := OvcVirtualListbox1.ItemIndex;
  if I > -1 then
    Label2.Caption := Format('Selected item: %d', [I])
  else
    Label2.Caption := '(none)';
end;

procedure TForm1.OvcVirtualListbox1IsSelected(Sender: TObject;
  Index: Longint; var Selected: Boolean);
begin
  if Index <= 500 then
    Selected := Select[Index]
  else
    Selected := False;
end;

procedure TForm1.OvcVirtualListbox1Select(Sender: TObject; Index: Longint;
  Selected: Boolean);
begin
  if Index = -1 then
    FillChar(Select, SizeOf(Select), Selected)
  else if Index <= 500 then
    Select[Index] := Selected;
end;

procedure TForm1.OvcVirtualListbox1TopIndexChanged(Sender: TObject;
  NewTopIndex: Longint);
begin
  Label3.Caption := Format('Top index: %d', [NewTopIndex])
end;

procedure TForm1.OvcVirtualListBox2GetItem(Sender: TObject; Index: Integer;
  var ItemString: String);
begin
  ItemString := IntToStr(Index);
end;

procedure TForm1.btnCloseClick(Sender: TObject);
begin
  Close;
end;

end.

File: examples/CBuildr4/exvlbu.cpp
//---------------------------------------------------------------------------
#include <vcl\vcl.h>
#pragma hdrstop

#include "ExVlbU.h"
//---------------------------------------------------------------------------
#pragma link "OvcVlb"
#pragma link "OvcBase"
#pragma link "OvcVLB"
#pragma resource "*.dfm"
TForm1 *Form1;
//---------------------------------------------------------------------------
__fastcall TForm1::TForm1(TComponent* Owner)
  : TForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TForm1::OvcVirtualListbox1GetItem(TObject *Sender, int Index,
  AnsiString &ItemString)
{
  // cheat and just return the index number as the item string
  ItemString = "Item # " + String((int)Index);
}
//---------------------------------------------------------------------------
void __fastcall TForm1::OvcVirtualListbox1GetItemColor(TObject *Sender,
  int Index, TColor &FG, TColor &BG)
{
  if (Index != OvcVirtualListbox1->ItemIndex) {
    if (Index % 13 == 0)
      FG = clRed;
    else if (Index % 7 == 0)
      FG = clGreen;
    else if (Index % 2 == 1)
      FG = clBlue;
  }
}
//---------------------------------------------------------------------------
void __fastcall TForm1::OvcVirtualListbox1Click(TObject *Sender)
{
  int i = OvcVirtualListbox1->ItemIndex;
  if (i > -1)
    Label2->Caption = "Selected item: " + String((int)i);
  else
    Label2->Caption = "(none)";
}
//---------------------------------------------------------------------------
void __fastcall TForm1::OvcVirtualListbox1IsSelected(TObject *Sender,
  int Index, bool &Selected)
{
  if (Index <= 500)
    Selected = Select[Index];
  else
    Selected = false;
}
//---------------------------------------------------------------------------
void __fastcall TForm1::OvcVirtualListbox1Select(TObject *Sender, int Index,
  bool Selected)
{
  if (Index == -1)
    for (int i=0;i<500;i++)
      Select[i] = Selected;
  else if (Index <= 500)
    Select[Index] = Selected;
}
//---------------------------------------------------------------------------
void __fastcall TForm1::OvcVirtualListbox1TopIndexChanged(TObject *Sender,
  int NewTopIndex)
{
  Label3->Caption = "Top index: " + String((int)NewTopIndex);
}
//---------------------------------------------------------------------------
void __fastcall TForm1::BitBtn1Click(TObject *Sender)
{
  Close();    
}
//---------------------------------------------------------------------------

